### PR TITLE
GLT-2944 Inhibit duplicate-index actions on Pool Orders

### PR DESCRIPTION
**Pull request by avarsava**
* Check on adding Library Aliquots (& therefore, Pool Order creation) for duplicate & near duplicate indices, block action
* Check if pool is known to contain duplicate or near duplicate indices on linking to pool order, block action

### CODE REVIEW

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java [hunk 1]
```diff
@@ -140,6 +150,31 @@ protected void collectValidationErrors(PoolOrder object, PoolOrder beforeChange,
     if (!object.isDraft() && object.getOrderLibraryAliquots().isEmpty()) {
       errors.add(new ValidationError("Non-draft order must include at least one library aliquot"));
     }
+
+    if(strictPools) validateIndices(object, beforeChange, errors);
+  }
+
+  private void validateIndices(PoolOrder object, PoolOrder beforeChange, List<ValidationError> errors){
```
Comment (callunity) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java]: What do you think about renaming this to something like `validateNoAdditionalDuplicateIndices`? (I'm imagining a future scenario where we might want to do additional validation on indices in all cases, not just in cases where `strictPools` is true)

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java [hunk 2]
```diff
@@ -140,6 +150,31 @@ protected void collectValidationErrors(PoolOrder object, PoolOrder beforeChange,
...
+    // Work based on whether bad index count increases, rather than >0, in case Pool Orders already exist w >1
+    if(indexChecker.getDuplicateIndicesSequences(beforeChange).size()
+            < indexChecker.getDuplicateIndicesSequences(object).size()
+            || indexChecker.getNearDuplicateIndicesSequences(beforeChange).size()
+            < indexChecker.getNearDuplicateIndicesSequences(object).size()) {
```
[the remark this answers is not in the document]
Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java]: I'm checking whether the number of duplicate or near-duplicate indices will increase with the current change; as far as I understand it, it doesn't matter what the sequences themselves are

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java (hunk 1 above)
Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java]: Done!

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (callunity) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java]: I guess I was thinking of the case where the lab removes one set of duplicates and replaces it with another... but looking at how many pool orders we actually have (one in staging) it's vanishingly unlikely that they'll have existing bad orders that they'll edit to also be bad.

[the remark this answers is not in the document]
Comment (avarsava) [reply miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java]: I believe removing one duplicate and replacing it with another would also count as 2 changes, so they'd be stopped from doing the second

**Pull request merged**